### PR TITLE
Node.js requirement

### PR DESCRIPTION
**Pull request by henryju**
Better support for Node.js and JRE requirements

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java [hunk 1]
```diff
@@ -24,25 +24,96 @@
...
+import org.sonarsource.sonarlint.core.client.api.common.PluginDetails;
+import org.sonarsource.sonarlint.core.client.api.common.SkipReason;
+import org.sonarsource.sonarlint.core.client.api.common.analysis.AnalysisResults;
 
 public class AnalysisRequirementNotifications {
 
-  private static final NotificationGroup ANALYSIS_PROBLEM_GROUP = NotificationGroup.balloonGroup("SonarLint: Analysis Requirement");
+  private static final NotificationGroup ANALYZER_REQUIREMENT_GROUP = NotificationGroup.balloonGroup("SonarLint: Analyzer Requirement");
+
+  private static final Set<String> alreadyNotified = new HashSet<>();
```
Comment (jblievremont) [inline src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java:51]: This means that users will have to reload SonarLint / restart the IDE to have the checks performed again, right?

Let's imagine someone has the warning about Node, installs it while the IDE is running (e.g with a package manager that puts it on the `$PATH`); then technically the requirement is met, but SonarLint will only notice it on restart.

I believe we have to either:
- mention that the IDE must be restarted in the message about Node
- make sure to clean this cache when the path to Node is updated in the settings

File: src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java [hunk 2]
```diff
@@ -24,25 +24,96 @@
...
+              runtimeRequirement.getMinVersion(), l.getLabel(), runtimeRequirement.getCurrentVersion());
+            createNotificationOnce(project, title, content, new NotificationListener.UrlOpeningListener(true));
+          } else if (runtimeRequirement.getRuntime() == SkipReason.UnsatisfiedRuntimeRequirement.RuntimeRequirement.NODEJS) {
+            StringBuilder content = new StringBuilder(String.format("SonarLint requires Node.js runtime version %s or later to analyze %s code.", runtimeRequirement.getMinVersion(), l.getLabel()));
+            if (runtimeRequirement.getCurrentVersion() != null) {
+              content.append(String.format(" Current version is %s.", runtimeRequirement.getCurrentVersion()));
+            }
+            content.append("<br>Please configure the Node.js path in the <a href=\"#edit-settings\">SonarLint settings</a>.");
```
Comment (jblievremont) [inline src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java:74]: I guess the new settings will come in another PR?

File: src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java [hunk 3]
```diff
@@ -68,14 +68,14 @@ public synchronized SonarLintFacade getFacade(boolean logDetails) throws Invalid
...
     SonarLintConsole console = SonarLintUtils.getService(myProject, SonarLintConsole.class);
     if (projectSettings.isBindingEnabled()) {
-      String serverId = projectSettings.getServerId();
+      String connectionId = projectSettings.getServerId();
       String projectKey = projectSettings.getProjectKey();
-      checkBindingStatus(notifications, serverId, projectKey);
+      checkBindingStatus(notifications, connectionId, projectKey);
       if (logDetails) {
-        console.info(String.format("Using configuration of '%s' in server '%s'", projectKey, serverId));
+        console.info(String.format("Using connection '%s' for project '%s'", connectionId, projectKey));
```
Comment (jblievremont) [inline src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java:75]: :+1:

**Review by jblievremont — APPROVED** [review]
Looks good on the functional side, see my comments about the "cache" of checks + settings declaration.

File: src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java (hunk 1 above)
Comment (henryju) [reply src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java:51]: Cache cleaned in the next commit, when the setting is updated

File: src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java (hunk 2 above)
Comment (henryju) [reply src/main/java/org/sonarlint/intellij/core/AnalysisRequirementNotifications.java:78]: in fact I added it to the same PR, sorry if it makes the review more difficult

**Comment by henryju** [comment]
@jblievremont there is the latest changes added in the last commits. To ease testing you can update the embedded SonarJS plugin to 6.5.0.13345 (thanks to https://github.com/SonarSource/SonarJS/pull/2172)

**Review by jblievremont — APPROVED** [review]
LGTM :+1:

**Pull request merged**